### ISSUE TITLE
Bundler doesn’t honor --user-install

### ISSUE THREAD

**Issue by Quintus** [bundler/bundler#2565]
Hi there,

I have this in my `/etc/gemrc`:

```
install: --user-install
update: --user-install
```

which works fine when installing gems directly with RubyGems (`gem install <gemname>`). However, when instructing bundler to install from a Gemfile, this doesn’t work and instead asks me to use `sudo` to install the gems system-wide, which is undesired.

```
% cat Gemfile 
source "https://rubygems.org"

gem "coderay"
(1010) [9:42:23 quintus@roma] /tmp/foo
% bundle install
Fetching gem metadata from https://rubygems.org/..
Resolving dependencies...
Enter your password to install the bundled RubyGems to your system: 

SystemExit: exit
An error occurred while installing coderay (1.0.9), and Bundler cannot continue.
Make sure that `gem install coderay -v '1.0.9'` succeeds before bundling.
(1011) [9:42:44 quintus@roma] /tmp/foo
% gem install coderay
Fetching: coderay-1.0.9.gem (100%)
Successfully installed coderay-1.0.9
invalid options: -f emerald, -SNw2
(invalid options are ignored)
Parsing documentation for coderay-1.0.9
Installing ri documentation for coderay-1.0.9
Done installing documentation for coderay after 3 seconds
1 gem installed
(1012) [9:43:01 quintus@roma] /tmp/foo
% gem which coderay
/home/quintus/.gem/ruby/2.0.0/gems/coderay-1.0.9/lib/coderay.rb
(1013) [9:43:06 quintus@roma] /tmp/foo
% 
```

Software in use:

```
(1017) [9:50:55 quintus@roma] /tmp/foo
% ruby -v
ruby 2.0.0p247 (2013-06-27 revision 41674) [x86_64-linux]
(1018) [9:50:58 quintus@roma] /tmp/foo
% gem -v
2.0.5
(1019) [9:50:59 quintus@roma] /tmp/foo
% bundle -v
Bundler version 1.3.5
(1020) [9:51:01 quintus@roma] /tmp/foo
% 
```

Valete,
Marvin

**Comment by omninonsense** [comment]
I'd like to resurrect this issue! Getting this to work would be as simple as loading `.gemrc` (it's a YAML file), and looking if `gem`, `install`, `update` or `uninstall` keys have the `--[no-]user-install` flag in the string, and respect it.

Now that I think of it, it'd be nice if other options were supported as well, and let users opt-out with `--[no-]gemrc` when running bundler,  and also override those [`.gemrc`] options when using bundler cmd args.

**Comment by indirect** [comment]
Bundler doesn't even have a concept of "user install". What would this setting activate?

**Comment by omninonsense** [comment]
> Bundler doesn't even have a concept of "user install". What would this setting activate?

I know *Bundler* doesn't, but gems do, or at least *gem, the command* does. It would be similar to `--path $HOME/.gem`&mdash;maybe it would even behave identically. IIRC, you can even get that path using `Gem.user_dir` (maybe it's `userdir`; can't test it right now).

The idea is that Bundler is used to manage gems, so one would expect it respects `.gemrc`.

**Comment by indirect** [comment]
The `gem` command manages a single pool of gems used by all Ruby processes, while Bundler manages separate pools of gems, one per project or application. Even though both install gems, I am not aware of any options from the `.gemrc` file that work as a direct mapping to Bundler.

If you want to relocate the `bundle install --system` directory to `$HOME/~.gem`, you can just set `$GEM_HOME` like `chruby` does. The `$GEM_HOME` environment variable is respected by both Bundler and RubyGems.

**Comment by omninonsense** [comment]
> The gem command manages a single pool of gems used by all Ruby processes, while Bundler manages separate pools of gems, one per project or application. Even though both install gems, I am not aware of any options from the .gemrc file that work as a direct mapping to Bundler.

Agreed. However, by default Bundler installs gems globally (it even elevates using `sudo` to accomplish this). The problem is that `bundle install --system` is to be the default behaviour when running `bundle install` (without `$BUNDLE_PATH` or `$GEM_HOME` being set). IHMO, the ideal default behaviour would be `--path .bundle/`, but I suppose that's too much to ask for since people are already used to bundler installing gems system-wide.

> If you want to relocate the bundle install --system directory to $HOME/~.gem, you can just set $GEM_HOME like chruby does. The $GEM_HOME environment variable is respected by both Bundler and RubyGems.

I know it is, but it's a bit of a hacky solution. I believe that:

1. changing default behaviour (unlikely) to gems in something like `./.bundle`,
2. or respecting `.gemrc`, 

instead of implicitly installing gems system-wide using `sudo` (which I think is a horrendous default) would be more desirable.

The current behaviour is *fine*, but I could be *better*.

**Comment by indirect** [comment]
I completely agree that it could be better, and Bundler 2.0 will default to --path .bundle for that reason. :)

**Comment by omninonsense** [comment]
> I completely agree that it could be better, and Bundler 2.0 will default to --path .bundle for that reason. :)

Oh, I wasn't aware of that. This is *amazing* news! Thanks for being so patient with me about this! :-)

Cheers!

**Comment by noraj1337** [comment]
I know necrobumping is bad but I think about those who will end here like me:


Add these lines at the end of your `~/.zshrc` or `~/.bashrc`
```
#Setting the GEM_PATH and GEM_HOME variables may not be necessary, check 'gem env' output to verify whether both variables already exist 
export GEM_HOME=$(ls -t -U | ruby -e 'puts Gem.user_dir')
export GEM_PATH=$GEM_HOME
export PATH=$PATH:$GEM_HOME/bin
```

So all what your user install with gem is installed under his own home directory `/home/user/.gem/ruby/2.4.0/bin`. `bundle info [gemname]` can show you were the gem is installed and `gem env` show you the whole RubyGems Environment.

PS : `/etc/gemrc` with `gem: --user-install`. This is the [archlinux way](https://wiki.archlinux.org/index.php/Ruby).

**Comment by akostadinov** [comment]
@noraj , thanks a lot for the idea. I just upgraded and I found `bundle` trying to install with `sudo` which is very unreasonable for me. So far it has been installing to the users home `~/.gem/ruby`. Now it tries to either install with `sudo` or in a project vendor directory.

So what I did is to put in my bash profile:
```
export BUNDLE_PATH=$(ls -t -U | ruby -e 'puts Gem.user_dir')
export PATH="$PATH:$BUNDLE_PATH/bin"
```

Now both `gem` and `bundle` install to the same user-wide location. The only issue is that `gem` installs executables in `~/bin` while `bundle` installs them in `$BUNDLE_PATH/bin`. But this is bearable.
